types/src/data.rs

```diff
@@ -186,6 +187,120 @@ pub trait ProposalType:
     fn get_view_number(&self) -> <Self::NodeType as NodeType>::Time;
 }
 
+/// A state change encoded in a leaf.
+///
+/// [`DeltasType`] represents a [block](NodeType::BlockType), but it may not contain the block in
+/// full. It is guaranteed to contain, at least, a cryptographic commitment to the block, and it
+/// provides an interface for resolving the commitment to a full block if the full block is
+/// available.
+pub trait DeltasType<Block: Committable>:
+    Clone + Debug + for<'a> Deserialize<'a> + PartialEq + Send + Serialize + Sync
+{
+    /// Errors reported by this type.
+    type Error: std::error::Error;
+
+    /// Get a cryptographic commitment to the block represented by this delta.
+    fn block_commitment(&self) -> Commitment<Block>;
+
+    /// Get the full block if it is available, otherwise return this object unchanged.
+    ///
+    /// # Errors
+    ///
+    /// Returns the original [`DeltasType`], unchanged, in an [`Err`] variant in the case where the
+    /// full block is not currently available.
+    fn try_resolve(self) -> Result<Block, Self>;
```

Comment:
Yeah I meant this less as a polling mechanism and more as a status check on an immutable snapshot of a leaf.

I envision a separate operation on the `HotShotHandle` like
```rust
async fn resolve_block(deltas: impl DeltasType<BlockType>) -> Result<BlockType, Error>;
```
I'm not sure about putting this method on the `DeltasType` itself because then, as you hinted at, the `DeltasType` would need to have some internal state which is shared with some background task in HotShot (like a channel) and that doesn't serialize well. Really all you should need is a block commitment and some sort of handle to consensus and you should be able to get a block.

But I intentionally left this interface out because we shouldn't need it yet (because of the static committee) and we need to put some actual thought into what this should look like.